src/validators/dict.rs

```diff
@@ -147,4 +150,78 @@ macro_rules! build_validate {
 impl DictValidator {
     build_validate!(validate_dict, PyDict);
     build_validate!(validate_json_object, JsonObject);
+
+    fn validate_mapping<'s, 'data>(
+        &'s self,
+        py: Python<'data>,
+        input: &'data impl Input<'data>,
+        dict: &'data PyMapping,
+        extra: &Extra,
+        slots: &'data [CombinedValidator],
+        recursion_guard: &'s mut RecursionGuard,
+    ) -> ValResult<'data, PyObject> {
+        let output = PyDict::new(py);
+        let mut errors: Vec<ValLineError> = Vec::new();
+
+        let key_validator = self.key_validator.as_ref();
+        let value_validator = self.value_validator.as_ref();
+
+        let items = match dict.items() {
+            Ok(items) => items,
+            Err(err) => return Err(ValError::new(ErrorType::MappingType { error: err.to_string() }, input)),
+        };
+        for elem in items.iter()? {
+            let elem_t = elem.unwrap().downcast::<PyTuple>()?;
+            if elem_t.len() != 2 {
+                errors.push(ValLineError::new(
+                    ErrorType::MappingType {
+                        error: format!(
+                            "ValueError: expected tuple of length 2, but got tuple of length {}",
+                            elem_t.len()
+                        ),
+                    },
+                    input,
+                ));
+                break;
+            }
+            let key = elem_t.get_item(0)?;
```

Comment:
I cannot use `get_item_unchecked` here, because the `pypy` CI said "This method does not exist.", and I guess we are using a different version of "PYO3" for `pypy`?

Reply:
you can probably use compile time flags like we do here

https://github.com/pydantic/pydantic-core/blob/463ed7aa83b08c43bc699e7e5380026551e38876/src/validators/typed_dict.rs#L431-L434

Reply:
Sounds good to me.